src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```diff
@@ -312,6 +391,205 @@ public void processElement(final ProcessContext context) throws Exception {
         }
     }
 
+    // This transform takes marked reads and generates the metrics from those tests.
+    private static final class GenerateMetricsTransform extends PTransform<PCollection<GATKRead>, PCollection<KV<String, DuplicationMetrics>>> {
+        private static final long serialVersionUID = 1l;
+
+        /**
+         *  The header for all reads processed.
+        */
+        private final PCollectionView<SAMFileHeader> header;
+
+        public GenerateMetricsTransform(final PCollectionView<SAMFileHeader> header) {
+            this.header = header;
+        }
+
+        @Override
+        public PCollection<KV<String, DuplicationMetrics>> apply(final PCollection<GATKRead> preads) {
+            return preads
+                .apply(generateMetricsByLibrary())
+                .setCoder(KvCoder.of(StringUtf8Coder.of(), new DuplicationMetricsCoder()))
+                // Combine metrics for reads from the same library.
+                .apply(Combine.<String, DuplicationMetrics>perKey(new CombineMetricsFn()))
+                // For each library, finalize the metrics by computing derived metrics and dividing paired counters
+                // by 2.
+                .apply(finalizeMetrics());
+        }
+
+        private PTransform<PCollection<? extends GATKRead>, PCollection<KV<String, DuplicationMetrics>>> generateMetricsByLibrary() {
+            return ParDo
+                .named("generate metrics by library")
+                .withSideInputs(header)
+                .of(new DoFn<GATKRead, KV<String, DuplicationMetrics>>() {
+                        private static final long serialVersionUID = 1l;
+
+                        @Override
+                        public void processElement(final ProcessContext context) throws Exception {
+                            final GATKRead record = context.element();
+                            final SAMFileHeader h = context.sideInput(header);
+                            final String library = LibraryIdGenerator.getLibraryName(h, record.getReadGroup());
+                            DuplicationMetrics metrics = new DuplicationMetrics();
+                            metrics.LIBRARY = library;
+                            // These need to be initialized so that the DuplicationMetrics can be serialized safely.
+                            metrics.PERCENT_DUPLICATION = 0.0;
+                            metrics.ESTIMATED_LIBRARY_SIZE = 0L;
+                            // We have already filtered out secondary and supplementary reads in the partition transform.
```

Comment:
I'm pretty sure the unmapped reads are filtered out by the partition function as well:

[https://github.com/broadinstitute/hellbender/pull/749/files#diff-05fb5a0326f8c0d57413f307dfbb4187R168]
```
                 public int partitionFor(final GATKRead read, final int n) {
                     return read.isSecondaryAlignment() || read.isSupplementaryAlignment() || read.isUnmapped() ? ReadsPartition.NOT_PRIMARY.ordinal() : ReadsPartition.PRIMARY.ordinal();
                 }
```